src/hotspot/share/runtime/lockStack.inline.hpp

```diff
@@ -61,32 +65,100 @@ inline void LockStack::push(oop o) {
   verify("pre-push");
   assert(oopDesc::is_oop(o), "must be");
   assert(!contains(o), "entries must be unique");
-  assert(can_push(), "must have room");
+  assert(!is_full(), "must have room");
   assert(_base[to_index(_top)] == nullptr, "expect zapped entry");
   _base[to_index(_top)] = o;
   _top += oopSize;
   verify("post-push");
 }
 
-inline void LockStack::remove(oop o) {
+inline oop LockStack::bottom() const {
+  assert(to_index(_top) > 0, "must contain an oop");
+  return _base[0];
+}
+
+inline bool LockStack::is_empty() const {
+  return to_index(_top) == 0;
+}
+
+inline bool LockStack::is_recursive(oop o) const {
+  if (!VM_Version::supports_recursive_lightweight_locking()) {
+    return false;
+  }
+
+  assert(contains(o), "entries must exist");
+  int end = to_index(_top);
+  for (int i = 1; i < end; i++) {
+    if (_base[i - 1] == o && _base[i] == o) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
```

Comment:
Maybe a more prominent comment about how you can only recursively enter a lock if the top is the same object?